From 68b229b5212f7143692c037aaea4c46bcb3cfb76 Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Date: Fri, 13 Sep 2024 15:20:20 +0200
Subject: [PATCH 692/841] fix(st-crypto): check RNG seed error after null data
 read

The Reference Manual recommends to always verify that RNG_DR is
different from zero. Because when it is the case a seed error can
occur between RNG_SR polling and RND_DR output reading (rare event).

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git e41b190dfe8eabe5bd84207c02004c3d6e15ea3c fix(st-crypto): check RNG seed error after null data read]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@foss.st.com>
Change-Id: I04e0154d8bdfcc375b60191c5141a22b6dd6075f
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/404511
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
Tested-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/crypto/stm32_rng.c | 57 ++++++++++++++++++++++++-----------
 1 file changed, 40 insertions(+), 17 deletions(-)

diff --git a/drivers/st/crypto/stm32_rng.c b/drivers/st/crypto/stm32_rng.c
index 4f005a69d..f917a896e 100644
--- a/drivers/st/crypto/stm32_rng.c
+++ b/drivers/st/crypto/stm32_rng.c
@@ -149,6 +149,32 @@ static int stm32_rng_enable(void)
 	return 0;
 }
 
+static int check_data_validity(void)
+{
+	int nb_tries = RNG_TIMEOUT_US / RNG_TIMEOUT_STEP_US;
+	uint32_t status = mmio_read_32(stm32_rng.base + RNG_SR);
+
+	/* Exit if data is ready without any seed error */
+	if ((status & (RNG_SR_SECS | RNG_SR_SEIS | RNG_SR_DRDY)) != RNG_SR_DRDY) {
+		do {
+
+			if ((status & (RNG_SR_SECS | RNG_SR_SEIS)) != 0U) {
+				seed_error_recovery();
+			}
+
+			udelay(RNG_TIMEOUT_STEP_US);
+			nb_tries--;
+			if (nb_tries == 0) {
+				return -ETIMEDOUT;
+			}
+
+			status = mmio_read_32(stm32_rng.base + RNG_SR);
+		} while ((status & RNG_SR_DRDY) == 0U);
+	}
+
+	return 0;
+}
+
 /*
  * stm32_rng_read - Read a number of random bytes from RNG
  * out: pointer to the output buffer
@@ -159,7 +185,6 @@ int stm32_rng_read(uint8_t *out, uint32_t size)
 {
 	uint8_t *buf = out;
 	size_t len = size;
-	int nb_tries;
 	uint32_t data32;
 	int rc = 0;
 	unsigned int count;
@@ -169,22 +194,10 @@ int stm32_rng_read(uint8_t *out, uint32_t size)
 	}
 
 	while (len != 0U) {
-		nb_tries = RNG_TIMEOUT_US / RNG_TIMEOUT_STEP_US;
-		do {
-			uint32_t status = mmio_read_32(stm32_rng.base + RNG_SR);
-
-			if ((status & (RNG_SR_SECS | RNG_SR_SEIS)) != 0U) {
-				seed_error_recovery();
-			}
-
-			udelay(RNG_TIMEOUT_STEP_US);
-			nb_tries--;
-			if (nb_tries == 0) {
-				rc = -ETIMEDOUT;
-				goto bail;
-			}
-		} while ((mmio_read_32(stm32_rng.base + RNG_SR) &
-			  RNG_SR_DRDY) == 0U);
+		rc = check_data_validity();
+		if (rc != 0) {
+			goto bail;
+		}
 
 		count = 4U;
 		while (len != 0U) {
@@ -193,6 +206,16 @@ int stm32_rng_read(uint8_t *out, uint32_t size)
 			}
 
 			data32 = mmio_read_32(stm32_rng.base + RNG_DR);
+
+			while (data32 == 0U) {
+				rc = check_data_validity();
+				if (rc != 0) {
+					goto bail;
+				}
+
+				data32 = mmio_read_32(stm32_rng.base + RNG_DR);
+			}
+
 			count--;
 
 			(void)memcpy(buf, (uint8_t *)&data32, MIN(len, sizeof(uint32_t)));
-- 
2.47.2

